cm31_ntt/src/ntt.rs

```diff
@@ -0,0 +1,1233 @@
+pub struct PrecomputedTwiddles {
+    pub small: Vec<CF>,  // for the base NTT of size NTT_BLOCK_SIZE_FOR_CACHE
+    pub full: Vec<CF>,   // for the higher levels of recursion in the radix-8 context
+    pub stride: Vec<CF>, // for stride 2 or stride 4
+}
+
+pub fn precompute_twiddles(n: usize) -> Result<PrecomputedTwiddles, String> {
+    assert!(is_power_of(n as u32, 2), "n must be a power of 2");
+
+    // Case where n is a power of 8
+    if is_power_of(n as u32, 8) {
+        let (precomp_small, precomp_full) = precomp_for_ntt_r8_hybrid_p(n);
+        return Ok(PrecomputedTwiddles {
+            small: precomp_small,
+            full: precomp_full,
+            stride: vec![],
+        });
+    }
+
+    // Case where n is of the form 8^k * 2
+    if is_power_of((n / 2) as u32, 8) {
+        let (precomp_small, precomp_full) = precomp_for_ntt_r8_hybrid_p(n / 2);
+        let precomp_s2 = precomp_s2(n);
+        return Ok(PrecomputedTwiddles {
+            small: precomp_small,
+            full: precomp_full,
+            stride: precomp_s2,
+        });
+    }
+
+    // Case where n is of the form 8^k * 4
+    if is_power_of((n / 4) as u32, 8) {
+        let (precomp_small, precomp_full) = precomp_for_ntt_r8_hybrid_p(n / 4);
+        let precomp_s4 = precomp_s4(n);
+        return Ok(PrecomputedTwiddles {
+            small: precomp_small,
+            full: precomp_full,
+            stride: precomp_s4,
+        });
+    }
+    
+
+    Err("n must be a power of 2, or of the form 8^k * 2 or 8^k * 4".to_string())
+}
+
+/// Performs a radix-8 NTT on the polynomial f.
+/// This is unoptimised as it does not use any precomputed twiddles.
+/// It is also slower than ntt_r8_hybrid_p() as it allocates Vecs at each level of recursion, and doing so at
+/// the lower levels (to handle 8^1 to 8^5) incurs significant memory allocation overhead.
+/// @param f The coefficients of the polynomial to be transformed.
+/// @param w The n-th root of unity, where n is the length of f.
+/// @return The transformed polynomial in evaluation form.
+pub fn ntt_r8_vec(f: &Vec<CF>, w: CF) -> Vec<CF> {
+    let n = f.len();
+    debug_assert!(n >= 8, "n must be at least 8");
+    debug_assert!(is_power_of(n as u32, 8), "n must be a power of 8");
+
+    fn do_ntt(f: &Vec<CF>, w: CF) -> Vec<CF> {
+        let n = f.len();
+        if n == 1 {
+            return f.clone();
+        }
+
+        let m = n / 8;
+        let w_pow_8 = w.pow(8);
+
+        // Partition f into 8 subarrays.
+        let mut a = vec![vec![CF::zero(); m]; 8];
+
+        let mut res = vec![CF::zero(); n];
+        unsafe {
+            for i in 0..m {
+                let i_8 = i * 8;
+                for j in 0..8 {
+                    a[j][i] = *f.get_unchecked(i_8 + j);
+                }
+            }
+
+            // Recurse
+            let mut ntt_a = Vec::with_capacity(8);
+            for i in 0..8 {
+                ntt_a.push(do_ntt(&a[i], w_pow_8));
+            }
```

Comment:
if this is perf critical we could play around with `MaybeUninit` but otherwise an array of empty vecs could do the trick

```suggestion
            let mut ntt_a = [Vec::new(); 8];
            for i in 0..8 {
                ntt_a[i] = do_ntt(&a[i], w_pow_8);
            }
```